content/blog/elastic-apm-plugin.md

```diff
@@ -0,0 +1,111 @@
+
+One of the tools we can use to improve the observability of our applications are application performance monitoring agents (or APM agents). The agents provide information of what is going on inside our applications. They generally do this by instrumenting our code, without necessarily needing any changes within our own code. Most providers of APM tooling have agents available for a wide range of programming languages, such as Java, .Net, Python, PHP, etc. Using the instrumentation the agents can record events, like HTTP requests, database queries and messaging events. 
+
+To record events on when for example a HTTP requests starts and ends, the agent will need to hook into the web framework of your choice. The agent provider generally will provide a list of supported technologies and frameworks. In the case of the HTTP request example the agent will dynamically add some code around the web framework to record when a HTTP request starts and ends.
+
+So what do you actually get by using these agents? In the example of Elastic APM you'll be able to see a timeline visualisation like below.
+
+{{<img src="https://www.elastic.co/guide/en/apm/guide/current/images/apm-distributed-tracing.png" class="img-fluid" title="Elastic APM dashboard" >}}
+
+We can see which HTTP request are fired and what database calls are being made. This information can be extremly useful when trying to understand what's happening within you application, for example when looking into a performance issue.
+
+### What if your framework is not supported?
+
+But what to do when your framework of choice is not supported by the agent. By not doing anything you might be missing out on valuable information, so is there anything we can do? Luckily there is! Most APM agents will offer some way of adding manual instrumentation in your code. This might be suitable if you want to add some instrumentation for a very specific use case. But what if for example your database framework is not supported? Adding manual instrumentation for each query is not an ideal solution.
+
+Today we'll be looking at an example were a Java application was using R2DBC (a database framework) and Elastic as the APM provider. The Elastic APM agent offers a plugin api so we can write our own instrumentation without needing to instrument each indiviual query. 
+
+### How does the APM agent work?
+
+Before we look into how we can write a plugin for our example case, let's dive into how the Elastic APM agent works. The Elastic APM agent uses Byte Buddy to instrument code, Byte Buddy can modify Java classes during runtime. This way the agent can change a class without recompiling it. What will typically happen is that the agent will add some code when a instrumented method is entered and exited. 
```

Comment:
As a full Java n00b I had never heard of Byte Buddy. Should we link to it?

Reply:
Removed all reference to Byte Buddy as the exact isn't that important anyway